{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6c0eaee2",
   "metadata": {},
   "source": [
    "# Variant Analysis using VCF files\n",
    "\n",
    "After downloading the VCF files and variants form chromosome 19, the coding sequence annotations on chromosome 19 are extacted with:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "23deddc9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('19', '59108550', '59109183')]\n",
      "[('19', '80839', '80840', '0.14'), ('19', '90973', '90974', '0.0037'), ('19', '91105', '91106', '0.96'), ('19', '93541', '93542', '0.01'), ('19', '93817', '93818', '0.0041')]\n"
     ]
    }
   ],
   "source": [
    "def extact_seq_annotations(file_path):\n",
    "    annotations = []\n",
    "    with open (file_path, \"r\") as f:\n",
    "        for line in f: \n",
    "            # splits line into fields\n",
    "            fields = line.strip().split(\"\\t\")\n",
    "            chrom, start, end, attributes, a, strand = fields\n",
    "\n",
    "            attr_dict = dict(item.split(\"=\") for item in attributes.split(\";\") if \"=\" in item) \n",
    "            transcript_id = attr_dict.get(\"transcript_id\")\n",
    "            # writes the transcript_id, start, and end to an output file\n",
    "            with open(\"data/processed/chr19_exons_annotated.txt\", \"a\") as out_f:\n",
    "                out_f.write(f\"{chrom}\\t{start}\\t{end}\\n\")\n",
    "\n",
    "        annotations.append((chrom, start, end))\n",
    "    return print(annotations[:5])  # Return the first 5 annotations to test\n",
    "    \n",
    "extact_seq_annotations(\"data/raw/chr19_exons.bed\")\n",
    "\n",
    "def extract_vcf_variants(file_path):\n",
    "    variants = []\n",
    "    with open(file_path, \"r\") as f:\n",
    "        for line in f:\n",
    "\n",
    "            fields = line.strip().split(\"\\t\")\n",
    "            chrom, start, end, attributes, a, strand = fields\n",
    "            attr_dict = dict(item.split(\"=\") for item in attributes.split(\";\") if \"=\" in item) \n",
    "            allele_freq = attr_dict.get(\"AF\")\n",
    "\n",
    "            with open(\"data/processed/chr19_variants_annotated.txt\", \"a\") as out_f:\n",
    "                out_f.write(f\"{chrom}\\t{start}\\t{end}\\t{allele_freq}\\n\")\n",
    "            variants.append((chrom, start, end, allele_freq))\n",
    "\n",
    "    return print(variants[:5])  # Return the first 5 variants to test\n",
    "\n",
    "extract_vcf_variants(\"data/raw/chr19_variants.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f3173e2",
   "metadata": {},
   "source": [
    "Sorted the annotation files in terminal using \n",
    "`sort -k1,1 -k2,2n -u chr19_exons_annotated.txt -o ch19_exons_annotated_sorted.txt`\n",
    "\n",
    "Now to intersect the mutations with CDS gene annotations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2482a544",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pybedtools import BedTool\n",
    "\n",
    "variants = BedTool(\"data/processed/chr19_variants_annotated.bed\")\n",
    "exons = BedTool(\"data/processed/ch19_exons_annotated_sorted.bed\")\n",
    "\n",
    "variants_in_exons = variants.intersect(exons)\n",
    "with open(\"data/processed/chr19_variants_in_exons.txt\", \"w\") as out_f:\n",
    "    for variant in variants_in_exons:\n",
    "        out_f.write(str(variant) + \"\\n\")\n",
    "\n",
    "print(variants_in_exons.head())\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bedtools",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
