{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6c0eaee2",
   "metadata": {},
   "source": [
    "# Variant Analysis using VCF files\n",
    "\n",
    "After downloading the VCF files and variants form chromosome 19, the coding sequence annotations on chromosome 19 are extacted with:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "23deddc9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('19', '59108550', '59109183')]\n",
      "[('19', '80839', '80840', '0.14'), ('19', '90973', '90974', '0.0037'), ('19', '91105', '91106', '0.96'), ('19', '93541', '93542', '0.01'), ('19', '93817', '93818', '0.0041')]\n"
     ]
    }
   ],
   "source": [
    "def extact_seq_annotations(file_path):\n",
    "    annotations = []\n",
    "    with open (file_path, \"r\") as f:\n",
    "        for line in f: \n",
    "            # splits line into fields\n",
    "            fields = line.strip().split(\"\\t\")\n",
    "            chrom, start, end, attributes, a, strand = fields\n",
    "\n",
    "            attr_dict = dict(item.split(\"=\") for item in attributes.split(\";\") if \"=\" in item) \n",
    "            transcript_id = attr_dict.get(\"transcript_id\")\n",
    "            # writes the transcript_id, start, and end to an output file\n",
    "            with open(\"data/processed/chr19_exons_annotated.txt\", \"a\") as out_f:\n",
    "                out_f.write(f\"{chrom}\\t{start}\\t{end}\\n\")\n",
    "\n",
    "        annotations.append((chrom, start, end))\n",
    "    return print(annotations[:5])  # Return the first 5 annotations to test\n",
    "    \n",
    "extact_seq_annotations(\"data/raw/chr19_exons.bed\")\n",
    "\n",
    "def extract_vcf_variants(file_path):\n",
    "    variants = []\n",
    "    with open(file_path, \"r\") as f:\n",
    "        for line in f:\n",
    "\n",
    "            fields = line.strip().split(\"\\t\")\n",
    "            chrom, start, end, attributes, a, strand = fields\n",
    "            attr_dict = dict(item.split(\"=\") for item in attributes.split(\";\") if \"=\" in item) \n",
    "            allele_freq = attr_dict.get(\"AF\")\n",
    "\n",
    "            with open(\"data/processed/chr19_variants_annotated.txt\", \"a\") as out_f:\n",
    "                out_f.write(f\"{chrom}\\t{start}\\t{end}\\t{allele_freq}\\n\")\n",
    "            variants.append((chrom, start, end, allele_freq))\n",
    "\n",
    "    return print(variants[:5])  # Return the first 5 variants to test\n",
    "\n",
    "extract_vcf_variants(\"data/raw/chr19_variants.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f3173e2",
   "metadata": {},
   "source": [
    "Sorted the annotation files in terminal using \n",
    "`sort -k1,1 -k2,2n -u chr19_exons_annotated.txt -o ch19_exons_annotated_sorted.txt`\n",
    "\n",
    "Now to intersect the mutations with CDS gene annotations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2482a544",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19\t105020\t105021\tRSQ=0.9089;LDAF=0.0692;AC=145;AA=G;AN=2184;VT=SNP;ERATE=0.0026;THETA=0.0125;SNPSOURCE=LOWCOV;AVGPOST=0.9851;AF=0.07;ASN_AF=0.03;AMR_AF=0.06;AFR_AF=0.09;EUR_AF=0.08\t.\t+\n",
      " 19\t159547\t159548\tERATE=0.0013;AVGPOST=0.9907;LDAF=0.0047;AN=2184;THETA=0.0171;VT=SNP;AA=.;AC=0;SNPSOURCE=LOWCOV;RSQ=0.0919;AF=0\t.\t+\n",
      " 19\t230129\t230130\tAA=GATC;AC=1311;AF=0.60;AFR_AF=0.59;AMR_AF=0.57;AN=2184;ASN_AF=0.88;AVGPOST=0.6811;ERATE=0.0090;EUR_AF=0.42;LDAF=0.5828;RSQ=0.4753;THETA=0.0030;VT=INDEL\t.\t+\n",
      " 19\t279506\t279507\tAN=2184;AC=8;THETA=0.0008;RSQ=0.9326;VT=SNP;LDAF=0.0038;SNPSOURCE=LOWCOV;ERATE=0.0003;AVGPOST=0.9995;AA=t;AF=0.0037;ASN_AF=0.01;EUR_AF=0.0040\t.\t+\n",
      " 19\t279508\t279509\tLDAF=0.0009;RSQ=0.6705;AA=g;AN=2184;VT=SNP;SNPSOURCE=LOWCOV;AC=1;THETA=0.0051;ERATE=0.0003;AVGPOST=0.9992;AF=0.0005;ASN_AF=0.0017\t.\t+\n",
      " 19\t279512\t279513\tAC=7;AN=2184;LDAF=0.0032;THETA=0.0008;VT=SNP;RSQ=0.9797;SNPSOURCE=LOWCOV;AA=a;ERATE=0.0003;AVGPOST=0.9999;AF=0.0032;AFR_AF=0.01\t.\t+\n",
      " 19\t279555\t279556\tAC=7;RSQ=0.6699;AN=2184;VT=SNP;AA=c;LDAF=0.0029;AVGPOST=0.9973;SNPSOURCE=LOWCOV;THETA=0.0007;ERATE=0.0003;AF=0.0032;AMR_AF=0.0028;AFR_AF=0.01\t.\t+\n",
      " 19\t279644\t279645\tERATE=0.0005;LDAF=0.1989;AN=2184;AVGPOST=0.9942;VT=SNP;THETA=0.0006;RSQ=0.9863;SNPSOURCE=LOWCOV;AC=435;AA=t;AF=0.20;ASN_AF=0.10;AMR_AF=0.23;AFR_AF=0.19;EUR_AF=0.27\t.\t+\n",
      " 19\t279667\t279668\tRSQ=0.7408;AVGPOST=0.9994;AA=g;AN=2184;VT=SNP;SNPSOURCE=LOWCOV;AC=1;THETA=0.0010;ERATE=0.0003;LDAF=0.0008;AF=0.0005;EUR_AF=0.0013\t.\t+\n",
      " 19\t279793\t279794\tLDAF=0.0024;THETA=0.0004;AN=2184;AVGPOST=0.9997;RSQ=0.9449;VT=SNP;AA=c;AC=5;SNPSOURCE=LOWCOV;ERATE=0.0003;AF=0.0023;AFR_AF=0.01\t.\t+\n",
      " None\n"
     ]
    }
   ],
   "source": [
    "from pybedtools import BedTool\n",
    "\n",
    "variants = BedTool(\"data/raw/chr19_variants.bed\")\n",
    "exons = BedTool(\"data/processed/ch19_exons_annotated_sorted.txt\")\n",
    "\n",
    "variants_in_exons = variants.intersect(exons)\n",
    "\n",
    "print(variants_in_exons.head())\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bedtools",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
