{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6c0eaee2",
   "metadata": {},
   "source": [
    "# Variant Analysis using VCF files\n",
    "\n",
    "After downloading the VCF files and variants form chromosome 19, the coding sequence annotations on chromosome 19 are extacted with:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23deddc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('19', ('59108550', '59109183'))]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def extact_seq_annotations(file_path):\n",
    "    annotations = {}\n",
    "    with open (file_path, \"r\") as f:\n",
    "        for line in f: \n",
    "            # splits line into fields\n",
    "            fields = line.strip().split(\"\\t\")\n",
    "            chrom, start, end, attributes, a, strand = fields\n",
    "\n",
    "            attr_dict = dict(item.split(\"=\") for item in attributes.split(\";\") if \"=\" in item) \n",
    "            transcript_id = attr_dict.get(\"transcript_id\")\n",
    "            # writes the transcript_id, start, and end to an output file\n",
    "            with open(\"data/processed/chr19_exons_annotated.txt\", \"a\") as out_f:\n",
    "                out_f.write(f\"{chrom}\\t{start}\\t{end}\\n\")\n",
    "\n",
    "            annotations[chrom] = (start, end)\n",
    "    return list(annotations.items())[:5]  # Return the first 5 annotations to test\n",
    "    \n",
    "extact_seq_annotations(\"data/raw/chr19_exons.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f3173e2",
   "metadata": {},
   "source": [
    "Sorted the annotation files in terminal using \n",
    "`sort -k1,1 -k2,2n chr19_exons_annotated.txt -o ch19_exons_annotated_sorted.txt`\n",
    "\n",
    "Now to intersect the mutations with CDS gene annotations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2482a544",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "None\n"
     ]
    }
   ],
   "source": [
    "from pybedtools import BedTool\n",
    "\n",
    "variants = BedTool(\"data/raw/chr19_variants.bed\")\n",
    "exons = BedTool(\"data/processed/ch19_exons_annotated_sorted.txt\")\n",
    "\n",
    "variants_in_exons = variants.intersect(exons)\n",
    "\n",
    "print(variants_in_exons.head())\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bedtools",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
